urlb534bf3fca_akademi4eg_aka_uvc_tgz-pJ8-net_dtkanov_blocks_core_AppCoreJ8/src/net/dtkanov/blocks/circuit/high_level/derived/ALU.java

```diff
@@ -0,0 +1,376 @@
+		}
+		initOperations();
+		initFlags();
+	}
+	
+	protected void initFlags() {
+		// Z-flag. Set if result is zero.
+		ORNode cmpz_out[] = new ORNode[bitness-1];
+		cmpz_out[0] = new ORNode();
+		cmpz_out[0].connectSrc(outMUXs[0], 0, 0);
+		cmpz_out[0].connectSrc(outMUXs[0], 1, 1);
+		for (int j = 1; j < bitness-1; j++) {
+			cmpz_out[j] = new ORNode();
+			cmpz_out[j].connectSrc(outMUXs[0], j+1, 0);
+			cmpz_out[j].connectSrc(cmpz_out[j-1], 0, 1);
+		}
+		NOTNode cmp_neg = new NOTNode();
+		cmp_neg.connectSrc(cmpz_out[bitness-2], 0, 0);
+		cmp_neg.connectDst(0, out_z_flag, 0);
+		
+		// S-flag. Set if result is negative.
+		out_s_flag.connectSrc(outMUXs[0], bitness-1, 0);
+		
+		// P-flag. Set if number of 1s in result is even, e.g. 0110.
+		// TODO. What is a proper value for zero result of operation?
+		XORNode ones_out[] = new XORNode[bitness-1];
+		ones_out[0] = new XORNode();
+		ones_out[0].connectSrc(outMUXs[0], 0, 0);
+		ones_out[0].connectSrc(outMUXs[0], 1, 1);
+		for (int j = 1; j < bitness-1; j++) {
+			ones_out[j] = new XORNode();
+			ones_out[j].connectSrc(outMUXs[0], j+1, 0);
+			ones_out[j].connectSrc(ones_out[j-1], 0, 1);
+		}
+		NOTNode ones_neg = new NOTNode();
+		ones_neg.connectSrc(ones_out[bitness-2], 0, 0);
+		ones_neg.connectDst(0, out_p_flag, 0);
+		
+		// C-flag. Set if it was set
+		// in last add/sub/rotate/shift/inc/dec ops. Zero otherwise.
+		Mux c_muxs[] = new Mux[7];
+		c_muxs[0] = new Mux();
+		c_muxs[0].connectSrc(sh_right, bitness, 0);
+		c_muxs[0].connectSrc(sh_left, bitness, 1);
+		c_muxs[0].connectSrc(opNOPs[3], 0, 2);
+		
+		c_muxs[1] = new Mux();
+		c_muxs[1].connectSrc(outMUXs[0], bitness-1, 0); // rotate right
+		c_muxs[1].connectSrc(outMUXs[0], 0, 1); // rotate left
+		c_muxs[1].connectSrc(opNOPs[3], 0, 2);
+		
+		c_muxs[2] = new Mux();
+		NOTNode subtr_neg = new NOTNode();
+		subtr_neg.connectSrc(subtr, bitness, 0);
+		c_muxs[2].connectSrc(subtr_neg, 0, 0);
+		c_muxs[2].connectSrc(adder, bitness, 1);
+		c_muxs[2].connectSrc(opNOPs[3], 0, 2);
+		
+		c_muxs[3] = new Mux();
+		c_muxs[3].connectSrc(inc_dec, bitness, 0);
+		c_muxs[3].connectSrc(inc, bitness, 1);
+		c_muxs[3].connectSrc(opNOPs[3], 0, 2);
+		
+		c_muxs[4] = new Mux();
+		c_muxs[4].connectSrc(c_muxs[1], bitness, 0);
+		c_muxs[4].connectSrc(c_muxs[0], bitness, 1);
+		c_muxs[4].connectSrc(opNOPs[2], 0, 2);
+		
+		c_muxs[5] = new Mux();
+		c_muxs[5].connectSrc(c_muxs[3], bitness, 0);
+		c_muxs[5].connectSrc(c_muxs[2], bitness, 1);
+		c_muxs[5].connectSrc(opNOPs[2], 0, 2);
+		
+		c_muxs[6] = new Mux();
+		c_muxs[6].connectSrc(c_muxs[4], bitness, 0);
+		c_muxs[6].connectSrc(c_muxs[5], bitness, 1);
+		c_muxs[6].connectSrc(opNOPs[1], 0, 2);
+		
+		XORNode is_c = new XORNode();
+		is_c.connectSrc(opNOPs[0], 0, 0);
+		is_c.connectSrc(opNOPs[1], 0, 1);
+		
+		ANDNode res_c = new ANDNode();
+		res_c.connectSrc(is_c, 0, 0);
+		res_c.connectSrc(c_muxs[6], 0, 1);
+		out_c_flag.connectSrc(res_c, 0, 0);
+	}
+	
+	protected void initOperations() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks such as on lines 176-182, lines 183-189, lines 190-196, lines 197-202, lines 203-211 into separate methods.